### DISCUSSION TITLE
How to use ”time“ in kernel

### DISCUSSION THREAD

**Discussion by lzhucheng** [Parcels-code/Parcels#1559] [Q&A]
hi， @erikvansebille
 I want to know the meaning of "time" in the below code and how to use time in my kernel. Can you give some advices？

`def AdvectionRK4(particle, fieldset, time):
    """Advection of particles using fourth-order Runge-Kutta integration.

    Function needs to be converted to Kernel object before execution.
    """
    (u1, v1) = fieldset.UV[particle]
    lon1, lat1 = (particle.lon + u1*.5*particle.dt, particle.lat + v1*.5*particle.dt)
    (u2, v2) = fieldset.UV[time + .5 * particle.dt, particle.depth, lat1, lon1, particle]
    lon2, lat2 = (particle.lon + u2*.5*particle.dt, particle.lat + v2*.5*particle.dt)
    (u3, v3) = fieldset.UV[time + .5 * particle.dt, particle.depth, lat2, lon2, particle]
    lon3, lat3 = (particle.lon + u3*particle.dt, particle.lat + v3*particle.dt)
    (u4, v4) = fieldset.UV[time + particle.dt, particle.depth, lat3, lon3, particle]
    particle.lon += (u1 + 2*u2 + 2*u3 + u4) / 6. * particle.dt
    particle.lat += (v1 + 2*v2 + 2*v3 + v4) / 6. * particle.dt`

**Comment by erikvansebille** [comment]
Hi @lzhucheng. Thanks for asking. I agree it may be a bit confusing what `time` is, but it is the 'clock' of the loop. It is in seconds, relative to the first day of the data of the fieldset.   So after a day of simulation, you can expect `time=86400`. This is independent of the calendar used because parcels internally uses seconds since start of fieldset.

Does that help?

**Reply by lzhucheng** [reply]
Thanks for your answer.
Does "time" influence the trajectory of particles? I got different results when I use (u1, v1) = fieldset.UV[particle] and (u1, v1) = fieldset.UV[time, particle.depth, particle.lat, particle.lon] in AdvectionRK4 kernel. 
so I want to know how to use "time" in field sample fieldset in a kernel. 
In addition, why do some use (u1, v1) = fieldset.UV[particle] while others use (u1, v1) = fieldset.UV[time, particle.depth, particle.lat, particle.lon] in AdvectionRK4 kernel？

**Reply by erikvansebille** [reply]
Oh surprising that you get different results; that shouldn't happen. Can you send some more details of your experiment? A minimal breaking example would be useful

The reason why sometimes we use `fieldset.UV[particle]` is because this is faster in `Scipy`-mode, see [this part of the JIT-vs-Scipy tutorial](https://docs.oceanparcels.org/en/latest/examples/tutorial_jit_vs_scipy.html#Further-digging-into-Scipy-mode:-adding-particle-keyword-to-Field-sampling). For JIT mode, it shouldn't make any difference...

**Comment by lzhucheng** [comment]
Thanks for your answer.
I wonder about that i got diffent result between particle.lon(next loop) and particle.lon_nextloop. I want to kwon what happen about that？
Here is the code
`
class CustomParticle(JITParticle):
    age = Variable('age', dtype=np.float32, initial=0.)
    
    lon_nextloop = Variable('lon_nextloop', dtype=np.float32, to_write=True, initial=attrgetter('lon'))
    lat_nextloop = Variable('lat_nextloop', dtype=np.float32, to_write=True, initial=attrgetter('lat'))
    time_nextloop = Variable('time_nextloop', dtype=np.float32, to_write=True, initial=0.)   
    
    # particle_state: 
    particle_state = Variable('particle_state', dtype=np.int32, initial=0.) 
    beached_count = Variable('beached_count', dtype=np.int32, initial=0.)

def Advection(particle, fieldset, time):
    
    particle.lon = particle.lon_nextloop
    particle.lat = particle.lat_nextloop
    
    (u1, v1) = fieldset.UV[time, particle.depth, particle.lat, particle.lon]
    # (u1, v1) = fieldset.UV[particle]
    lon1_adv, lat1_adv = (particle.lon + u1*.5*particle.dt, particle.lat + v1*.5*particle.dt)

    (u2, v2) = fieldset.UV[time + .5 * particle.dt, particle.depth, lat1_adv, lon1_adv]
    lon2_adv, lat2_adv = (particle.lon + u2*.5*particle.dt, particle.lat + v2*.5*particle.dt)

    (u3, v3) = fieldset.UV[time + .5 * particle.dt, particle.depth, lat2_adv, lon2_adv]
    lon3_adv, lat3_adv = (particle.lon + u3*particle.dt, particle.lat + v3*particle.dt)

    (u4, v4) = fieldset.UV[time + particle.dt, particle.depth, lat3_adv, lon3_adv]

    particle_dlon = (u1 + 2*u2 + 2*u3 + u4) / 6. * particle.dt
    particle_dlat = (v1 + 2*v2 + 2*v3 + v4) / 6. * particle.dt     
    particle.lon_nextloop = particle.lon + particle_dlon
    particle.lat_nextloop = particle.lat + particle_dlat 
`

**Reply by erikvansebille** [reply]
Please have a look at the [kernel loop tutorial](https://docs.oceanparcels.org/en/latest/examples/tutorial_kernelloop.html), which is exactly about this topic. But this is not the code that breaks the `time` (point above), right?
